### ISSUE TITLE
site matching multiple rules with delayed start with never start

### ISSUE THREAD

**Issue by matkoniecz** [proginosko/LeechBlockNG#173]
in my case: blocked by ruleset A (90 seconds delay) and ruleset B (60 seconds delay).

After opening it never starts, it is repeatedly blocked by ruleset A and then by ruleset B.

![Peek 2020-10-23 09-55](https://user-images.githubusercontent.com/899988/96971785-e5de9c00-1515-11eb-8590-2f3a973c8875.gif)

To avoid XY problem: I setup nonoveridable delayed limit for some websites in the morning but I want them accessible after going through a delay. This turns it into a complete block.

```
setName1=a
sites1=openstreetmap.org
times1=0000-2400
limitMins1=
limitPeriod1=3600
limitOffset1=
conjMode1=false
days1=127
blockURL1=delayed.html?$S&$U
applyFilter1=false
closeTab1=false
filterName1=grayscale
activeBlock1=true
countFocus1=true
delayFirst1=true
delaySecs1=3
reloadSecs1=
allowOverride1=false
prevOpts1=false
prevGenOpts1=false
prevAddons1=true
prevSupport1=true
showTimer1=false
allowKeywords1=false
sitesURL1=
regexpBlock1=
regexpAllow1=
ignoreHash1=true
setName2=b
sites2=openstreetmap.org
times2=0000-2100
limitMins2=
limitPeriod2=3600
limitOffset2=
conjMode2=false
days2=127
blockURL2=delayed.html?$S&$U
applyFilter2=false
closeTab2=false
filterName2=grayscale
activeBlock2=false
countFocus2=true
delayFirst2=true
delaySecs2=2
reloadSecs2=
allowOverride2=true
prevOpts2=false
prevGenOpts2=false
prevAddons2=true
prevSupport2=true
showTimer2=false
allowKeywords2=false
sitesURL2=
regexpBlock2=
regexpAllow2=
ignoreHash2=true
numSets=2
sync=false
theme=
oa=0
password=podciąganie zrobione
hpp=true
timerVisible=true
timerSize=1
timerLocation=0
timerBadge=true
orm=5
ora=1
orc=true
warnSecs=
warnImmediate=true
contextMenu=true
toolsMenu=true
matchSubdomains=false
saveSecs=10
processActiveTabs=false
accessCodeImage=false
autoExportSync=true
```

**Comment by matkoniecz** [comment]
I would expect that in such case only rule with largest delay would apply to delay OR that that every rule would apply once OR that total cumulative countdown would appear.

**Comment by punyidea** [comment]
This has happened to me with Firefox and Windows 10.

**Comment by CosineP** [comment]
This has been a problem for a long time that I just put up with, but it's now a blocker for my desired configuration. Giving a bump here. Huge fan, have used Leechblock for over 7 years, would be happy to contribute a PR if I could be pointed in the right direction.

(In particular, the type of configuration this makes extremely annoying is when you have a catch-all blockset along with specific blocksets.)

**Comment by proginosko** [comment]
Can you create a minimal problem case and [email](mailto:leechblockng@proginosko.com) it to me (exported options)?

**Comment by matkoniecz** [comment]
> Can you create a minimal problem case and [email](mailto:leechblockng@proginosko.com) it to me (exported options)?

is one in OP ( https://github.com/proginosko/LeechBlockNG/issues/173#issue-728000520 ) insufficient, or do you want it specifically mailed?

**Comment by matkoniecz** [comment]
(I solved it for myself by preprocessing rules in my LB config generator)

**Comment by proginosko** [comment]
> > Can you create a minimal problem case and [email](mailto:leechblockng@proginosko.com) it to me (exported options)?
> 
> is one in OP ( [#173 (comment)](https://github.com/proginosko/LeechBlockNG/issues/173#issue-728000520) ) insufficient, or do you want it specifically mailed?

Fair point. 🙂 That was actually directed at @CosineP, but what you provided is fine.

Now I'm wondering why you need to have two overlapping block sets for your usage case. Why not a delay of 5 seconds for 0000-2100 and a delay of 3 seconds for 2100-2400?

In other words, what would be the usage case for two consecutive delays?

**Comment by matkoniecz** [comment]
> In other words, what would be the usage case for two consecutive delays?

In my case it is actually triggered by setup where

- some filters get flat 60s delay at all times
- some filters get 60s delay, except evening entertainment time between 0000-1830 and 19:58-2400
- some filters get complete block - except on 0000-1830 and 19:58-2400
- some filters get complete block, no exceptions

but if something appears in both delay groups one it effectively gets into the third one as loop is infinite except evening entertainment time

if something appears in second and third then it is misleadingly displayed as delayed then it turns out to be completely blocked

if something appears in one of groups and full block then it first appears as delayed/partially blocked and once that should let it it gets listed as fully blocked anyway

the two previous cases are quite frustrating and few times lead me to disable LB (wait, I waited, that is unfair!!11!11), then proceed to waste time.

I gave example of simplified setup as it is less fragile for testing (not time-dependent) but I can also try to craft more real case that would still be reasonable to implement. Maybe:

- flat 1s delay
- 1s delay except one minute block at some 1min block of time in night
- block except one minute block at some 1min block of time in night
- full block

with some entries for each combination?

**Comment by matkoniecz** [comment]
So I guess that simple version of 

> In other words, what would be the usage case for two consecutive delays?

would be that they have some of other settings (timespan, overredibility) different.

Though the question would be which rule should be used: first matching one? Last matching one? Each matching one, but once time delay passed then it should not trigger again? Maybe all should try to pass at once? Maybe you should get warning that multiple rules apply when you enter site? Maybe window should be split on parts and you should see multiple block messages?

What if during delay some new rule start applying? What if some rule expires during delay and starts appearing again?
